Allow generation of new(!) individual entries

jonandernovella opened the pull request:
With this PR I added the backend and frontend bits needed for creating new individual identifiers that are not entered manually by an user.

-  It checks how many breeding events happened on a given year in a herd to figure out the "Kull". It assumes that this is enough information to assign the next number.
- The last number of the identifier is calculated based on the litter size of the breeding event.

jonandernovella committed:
Add method for next ind number

This should create a new valid number for an individual based on breeding events on that year of a given herd.

jonandernovella committed:
Make ind number optional in form

It can be either system generated or passed in the form.

jonandernovella committed:
Update data access tests

They should now reflect the use case when we want the system to generate ind numbers.

HenrikeW commented on app/utils/database.py:
```diff
@@ -400,6 +400,53 @@ class Meta:  # pylint: disable=too-few-public-methods
+
+            next_litter = len(events)
+
+            individuals = (
+                Individual.select(Individual).where(
+                    Individual.breeding == str(breeding_event)
+                )
+            ).execute()
```
for the first individual of a litter, this won't find anything and `individuals` will be empty/undefined/null or something, right? Will line 436 still work then and give a `litter_size` of 0 or will the `len(individuals)` lead to an error?

jonandernovella replied:
With not finding anything do you mean that the query will return 0 records? If so, yes, `litter_size ` will be 0.

HenrikeW replied:
alright, then it's fine :+1:

HenrikeW commented on app/utils/database.py:
```diff
@@ -400,6 +400,53 @@ class Meta:  # pylint: disable=too-few-public-methods
+    def next_individual_number(herd, birth_date, breeding_event):
+        """
+        Returns the next litter number for a given year and herd.
+        """
```
the comment and the function name sound kind of contrary to me :) what about this?
```suggestion
    def next_individual_number(herd, birth_date, breeding_event):
        """
        Returns the number for the next individual in a litter for a given year and herd.
        """
```

HenrikeW reviewed (approved):
Looks good I think. Good work :+1:  Just for clarification: this doesn't _create_ new individuals yet, when a breeding is created, right? It only _allows_ to do so, or did I miss anything?
One picky comment about a comment from me. And another one about something that might be a problem, or it might not (lack of detailed python knowledge here to determine that). 
Also, some print statements and outcommented code left that you might want to remove, for the sake of tidiness ;-)

jonandernovella committed:
Update app/utils/database.py

Co-authored-by: Henrike Wiemker <65461017+HenrikeW@users.noreply.github.com>

jonandernovella committed:
Add frontend call for creating inds

jonandernovella committed:
Add better validation in backend

Validate date objects' type and do not set None valued properties.

jonandernovella committed:
Make a few individual properties optional

jonandernovella committed:
Improve frontend error handling

jonandernovella committed:
Fix format error

jonandernovella committed:
Clean up backend additions

jonandernovella committed:
Remove not needed checks

jonandernovella committed:
Fix generated new ind numbers

The current herd should be checked just in case. Also, we check for the exception when there is no individuals belonging to the breeding event yet.

jonandernovella committed:
Remove unused line

jonandernovella committed:
Remove unused import

jonandernovella committed:
Use origin herd and relocate query scope

jonandernovella committed:
Reformat code
